### PR TITLE
osd: add has_manifest_chunk to count chunks in snapshot

### PR DESCRIPTION
**Pull request by myoungwon**
cls_has_chunk does not cover snapshotted manifest object.
This leads to unexpected behavior during chunk scrub.

Signed-off-by: Myoungwon Oh <myoungwon.oh@samsung.com>

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

File: src/osd/PrimaryLogPG.cc [hunk 1]
```diff
@@ -3320,6 +3320,37 @@ void PrimaryLogPG::cancel_manifest_ops(bool requeue, vector<ceph_tid_t> *tids)
   }
 }
 
+bool PrimaryLogPG::has_manifest_chunk(ObjectContextRef obc, std::string& fp_oid) 
+{
+  if (!obc->obs.oi.has_manifest()) {
+    return false;
+  }
```
Comment (athanatos) [inline src/osd/PrimaryLogPG.cc:3327]: Probably want to remove this check from cls_has_chunk as well.

**Comment by athanatos** [comment]
What is cls_has_chunk suppose to do?  Why do we want to return true if any snapshot has that chunk?  Do we use it to validate whether a particular chunk reference is live?  Wouldn't it actually have to return the *number* of references to a particular chunk (an object might reference the same chunk twice at different offsets within the object)?

**Comment by myoungwon** [comment]
cls_has_chunk is used for chunk scrub by ceph_dedup_tool in order to find out the reference mismatch. 
Assuming that there are manifest and chunk object in two tiers like:
base tier: manifest object A (has a chunk object CHUNKA)
low tier: chunk object CHUNKA (has a reference of the object A)

Due to the false-positive design on reference counting, 

> [From https://github.com/ceph/ceph/pull/24230] As you know, the key idea behind of reference counting is false-positive, which means (manifest object (no ref), chunk object(has ref)) can be possible instead of (manifest object (has ref), chunk 1(no ref)). So, if we want to add chunk-scrub, it would be enough to add scrub operation in the chunk pool.

what we should do here is to examine that CHUNKA exists in the low tier, and then check that the manifest object A contains CHUNKA's reference. If not, we should remove the CHUNKA.

**Comment by athanatos** [comment]
Ok, what about the multiple reference case I mention above?

**Comment by myoungwon** [comment]
As your comment, there is a case as below,
Due to the same chunk in two different offsets within the object A, CHUNKA has two references of object A.
But, the object A does not be updated correctly.
Base tier: manifest object A (has chunk object CHUNKA )
Low tier: chunk object CHUNKA (has a reference of the object A and the object A)

In this case, cl_has_chunk() should return the number of references to decrease the reference count in CHUNKA.
Did you point out this case?

**Comment by athanatos** [comment]
Yeah, that's what I'm asking about.  Do we record how many references each backref holds?

**Comment by myoungwon** [comment]
Done. Can you take a look?

File: src/osd/PrimaryLogPG.cc [hunk 2]
```diff
@@ -3320,6 +3320,46 @@ void PrimaryLogPG::cancel_manifest_ops(bool requeue, vector<ceph_tid_t> *tids)
   }
 }
 
+int PrimaryLogPG::has_manifest_chunk(ObjectContextRef obc, std::string& fp_oid)
```
Comment (athanatos) [inline src/osd/PrimaryLogPG.cc:3323]: get_manifest_ref_count

File: src/osd/PrimaryLogPG.cc [hunk 3]
```diff
@@ -3320,6 +3320,46 @@ void PrimaryLogPG::cancel_manifest_ops(bool requeue, vector<ceph_tid_t> *tids)
...
+    hobject_t clone_oid = obc->obs.oi.soid;
+    clone_oid.snap = *p;
+    ObjectContextRef clone_obc = get_object_context(clone_oid, false);
+    if (!clone_obc) {
+      break;
+    }
+    get_adjacent_clones(clone_obc, obc_l, obc_g);
+    clone_obc->obs.oi.manifest.calc_refs_to_inc_on_set(
+      obc_g ? &(obc_g->obs.oi.manifest) : nullptr ,
+      nullptr,
```
Comment (athanatos) [inline src/osd/PrimaryLogPG.cc:3349]: Missing obc_l?

I'm not sure this is actually correct, probably you want to start at the oldest clone and work forward to head populating only obc_l.

Comment (myoungwon) [reply src/osd/PrimaryLogPG.cc:3349]: Does this provide the same output as our intention?
This commit counts all chunks in the head object and then checks the latest clone in reverse order.
(head -> snap[0] -> snap[1] in the below example)

 // foo snap[1]:      [HI] [HI]
 // foo snap[0]: [er] [ai] [SI]
 // foo head   : [er] [bi] [SI]

Comment (athanatos) [reply src/osd/PrimaryLogPG.cc:3349]: Hmm, I suppose it does.

File: src/osd/PrimaryLogPG.cc [hunk 4]
```diff
@@ -3320,6 +3320,46 @@ void PrimaryLogPG::cancel_manifest_ops(bool requeue, vector<ceph_tid_t> *tids)
...
+      refs);
+    if (!refs.is_empty()) {
```
Comment (athanatos) [inline src/osd/PrimaryLogPG.cc:3351]: You don't need to do this check, the for loop before would just immediately terminate.

**Comment by myoungwon** [comment]
@athanatos I addressed your comment.

**Review by athanatos — APPROVED** [review]

**Comment by athanatos** [comment]
Needs a rados suite run prior to merge.

**Comment by myoungwon** [comment]
I just rebased the commits to run a rados suite on top of the latest master.

https://pulpito.ceph.com/myoungwon-2021-01-21_09:53:20-rados-wip-manifest-chunk-scrub-distro-basic-smithi/
https://pulpito.ceph.com/myoungwon-2021-01-24_05:30:33-rados-wip-manifest-chunk-scrub-distro-basic-smithi/

**Comment by athanatos** [comment]
@myoungwon There are a ton of dead and failed runs there.

**Comment by myoungwon** [comment]
https://pulpito.ceph.com/myoungwon-2021-01-25_12:40:07-rados-wip-manifest-chunk-scrub-distro-basic-smithi/

Hm.. The dead and fail cases are related to this PR?
It seems that all dead cases did not finish their job due to test environment as below.

> 2021-01-24T06:06:15.487 ERROR:teuthology.orchestra.daemon.state:Failed to send signal 1: None
Traceback (most recent call last):
  File "/home/teuthworker/src/git.ceph.com_git_teuthology_master/teuthology/orchestra/daemon/state.py", line 108, in signal
    self.proc.stdin.write(struct.pack('!b', sig))
  File "/home/teuthworker/src/git.ceph.com_git_teuthology_master/virtualenv/lib/python3.6/site-packages/paramiko/file.py", line 405, in write
    self._write_all(data)
  File "/home/teuthworker/src/git.ceph.com_git_teuthology_master/virtualenv/lib/python3.6/site-packages/paramiko/file.py", line 522, in _write_all
    count = self._write(data)
  File "/home/teuthworker/src/git.ceph.com_git_teuthology_master/virtualenv/lib/python3.6/site-packages/paramiko/channel.py", line 1364, in _write
    self.channel.sendall(data)
  File "/home/teuthworker/src/git.ceph.com_git_teuthology_master/virtualenv/lib/python3.6/site-packages/paramiko/channel.py", line 846, in sendall
    sent = self.send(s)
  File "/home/teuthworker/src/git.ceph.com_git_teuthology_master/virtualenv/lib/python3.6/site-packages/paramiko/channel.py", line 801, in send
    return self._send(s, m)
  File "/home/teuthworker/src/git.ceph.com_git_teuthology_master/virtualenv/lib/python3.6/site-packages/paramiko/channel.py", line 1198, in _send
    raise socket.error("Socket is closed")
OSError: Socket is closed
>

In addition, the failed case is already tracked by https://tracker.ceph.com/issues/48786, https://tracker.ceph.com/issues/48915.
Another failed case caused by _wctx_finish in bluestore.

**Comment by tchaikov** [comment]
- https://pulpito.ceph.com/kchai-2021-01-27_08:19:23-rados-wip-kefu-testing-2021-01-27-1353-distro-basic-smithi/
- https://pulpito.ceph.com/kchai-2021-01-28_03:28:19-rados-wip-kefu-testing-2021-01-27-1353-distro-basic-smithi/

**Pull request merged**